knowledge-base/implementation-documentation/api-improvements.md

```diff
@@ -0,0 +1,209 @@
+    )
+  )
+```
+### Issues
+
+The inner filter is quite confusing, because `filter` implies that we are filtering the Address here, while it is actually filtering the business partner.
+
+### Alternatives
+
+`filterBy`
+
+```ts
+BusinessPartner.requestBuilder().getAll()
+  .filter(
+    BusinessPartner.TO_BUSINESS_PARTNER_ADDRESS.filterBy(
+      BusinessPartnerAddress.CITY.equals('Berlin')
+    )
+  )
+```
+
+flatten filters
+
+```ts
+BusinessPartner.requestBuilder().getAll()
+  .filter(
+    BusinessPartner.TO_SINGLE_LINK.CITY.equals('Berlin'),
+    BusinessPartner.TO_SINGLE_LINK.STATE.equals('Berlin'),
+    any(BusinessPartner.TO_MULTI_LINK.CITY.equals('Berlin'))
+  )
+```
+Concerns:
+* Matter of taste - not clearly better.
+* Longer for "deep" paths.
+* Not straight forward to implement. (Further investigation needed.)
+
+## Seperate Entity(Set) API and entity
+
+### Current API
+```ts
+const buPa: BusinessPartner = await BusinessPartner.requestBuilder().getAll().execute(destination);
+```
+
+### Issues
+The entity is a god-like element here. Separating could clean up.
+
+### Alternatives
+
+Split entity and api
+```ts
+const buPa: BusinessPartner = await BusinessPartnerApi.getAll()
+  .filter(BusinessPartner.FIRST_NAME.equals('Peter'))
+  .execute(destination);
+```
+* Could fix circular dependencies.
+* Could be implemented as a non-breaking change initially.
+* Realistic for 2.0.
+
+
+Split entity instance, api, and "helpers" (schema)
+```ts
+const buPa: BusinessPartner = await BusinessPartnerApi.getAll()
+  .filter(BusinessPartnerSchema.FIRST_NAME.equals('Peter'))
+  .execute(destination);
+```
+* Cleaner separation between responsibilities.
+* Importing three things is quite complex
+* Could be implemented as a non-breaking change initially.
+* Optionally call BusinessPartner => BusinessPartnerEntity.
+* Realistic for 2.0, but only useful if we consider function style filtering (at any point).
+
+## Function style filtering
+
+### Current API
+```ts
+BusinessPartner.requestBuilder().getAll()
+  .filter(BusinessPartner.FIRST_NAME.equals('Peter'))
+```
+
+### Issues
+This is more a matter of taste.
+To get closer to the look and feel of native JavaScript, instead of passing filters to `.filter` it might be nicer to allow functions.
+
+### Alternatives
+
+Using a function:
+```ts
+BusinessPartner.requestBuilder().getAll()
+  .filter(bupa => bupa.FIRST_NAME.equals('Peter'))
+```
+* We are only using static variables here. This is not clear anymore and might be more confusing than before.
+* Makes more sense when the schema is split from entity and api.
+* The examples below could be non-breaking changes done sequentially over time.
+
+Futher examples (and options):
+```ts
+BusinessPartner.requestBuilder().getAll()
+  .filter((bupa, { and }) => and(
+    bupa.FIRST_NAME.equals('Peter')
+  ))
+```
+
+```ts
+BusinessPartner.requestBuilder().getAll()
+  .filter((bupa, { and, equals }) => and(
+    equals(bupa.FIRST_NAME, 'Peter')
+  ))
+```
+
```

Comment:
```suggestion
As mentioned in the meeting I like the last approach because it would make all filter functions like equals, startsWith, greater,... symmetric.
```

Reply:
I think this can be seen as a pro for this approach, can you express it without "I" and commit it?